### ISSUE TITLE
Game building tools (master ticket)

### ISSUE THREAD

**Issue by ivan-mogilko** [adventuregamestudio/ags#2316]
This is a master ticket for tasks related to *decoupling game compilation from the Editor*. The final goal is to have **a minimal set of standalone tools enough for compiling a game data package** from the game project sources.
The previous ticket on this topic: #872, it contains a discussion and records first steps in this direction.
The general consensus was to write **standalone command-line programs in C++**, as that's something that may be compiled everywhere, and may reuse parts of the existing engine's code.
There have been a number of tools created since (see [Tools dir](https://github.com/adventuregamestudio/ags/tree/master/Tools) in our repo), pretty much enough for compiling the *game scripts*, but not covering all required tasks for creating game data. Several  tools more have to be done, and some of the existing ones may need adjustments.

Please note that this ticket is dedicated to writing tools, not replacing a build process in the Editor. That should be a separate task. In theory, replacement may be done in steps too, so long as that is convenient; but that's a topic for another discussion.

All the sub-tasks for this goal are assigned to a ["Minimal standalone tool set" milestone](https://github.com/adventuregamestudio/ags/issues?q=is%3Aopen+is%3Aissue+milestone%3A%22AGS+3%3A+Minimal+standalone+tool+set%22).
There's also a github "project" created, as an experiment for tracking the related tasks: 
https://github.com/adventuregamestudio/ags/projects/1
A wiki page I wrote, with a table of all the build steps, input and output data, which may help understanding the building process:
https://github.com/adventuregamestudio/ags/wiki/AGS-Game-Build-process-(3.5.*)

Above explains ags3 workflow, ags4 may have certain changes, primarily due to change in room format.

Following are lists of required tasks, separated into categories.

---

### Priority tools (required for both ags3 and ags4)

These are must have for building a game from a source.

- [ ] Build main game data from Game.agf (#1188). Must parse Game.agf, create game structs in memory, and write game28.dta.
- [ ] Export lists of game contents necessary for running script compiler and game packer (ticket TBD). Must parse Game.agf, and create simple text files with lists of stuff: scripts to compile, files to gather into package, etc. Something that may be used by any external script (bash, py, make, etc) which runs whole process.
- [ ] (OPTIONAL) Merge tools that export generated scripts from Game.agf (ticket TBD). These are tools named **agf2autoash** and **agf2glvar**, I did them two separate programs, but I think it's more convenient to have one that does one of or both things depending on command args. Because they both do thematically same thing: translate game contents to script as required for the scripts to compile.

### Secondary tools

These are tools that may be skipped when just building the game from "full" project state, but they are still required in order to get the project into this "full" state.

- [ ] Spritefile builder (ticket TBD). The purpose of the tool is to parse Game.agf for a list of sprites, generate sprites from their source files according to their import settings, and write a new spritefile (acsprset.spr) and optionally a sprite index (sprindex.dat).
Note that for this tool we need a image loading library in C++, as source files may come in various formats (BMP and PNG at minimum, but ideally everything that AGS Editor supported).
- [ ] Translation generator (ticket TBD). The purpose of the tool is to parse Game.agf, all scripts, and all room files (!), gather text strings, and create a translation source.
Note that AGS Editor has a concept of a "game text processor": an algorithm that knows which files and data to parse and how to retrieve text. Similar idea may be used when writing a standalone tool, because game text processing could be used for various purposes.
- [ ] Default config writer (ticket TBD). The purpose is to parse Game.agf for a component called "Default Setup" and generate default game config file (acsetup.cfg).

### Tools specific for ags4

- [ ] Room file compiler (ticket TBD). Must parse room's data.xml and compile room.crm out of it, packing all room assets together (backgrounds, masks, compiled room script). This has to be thought through first, as there may be few more *intermediate steps*. For example, room bgs and masks may be stored as PNGs, but has to be converted to raw bitmap data and compressed using LZW. Which means there has to be a PNG loading code in C++. Maybe we could make a separate dedicated tool for working with images, as we also might use one for compiling spritefile (see above); or maybe linking image-read/write functions to a room-compiling tool will be okay.
Also note that since 3.6.0 we do not have to package compiled room scripts inside the room.crm, and may place them as separate assets insides (roomN.o). I suggest to use this option, as that will simplify the room compiling tool.
- [ ] Translation compiler from PO source (ticket TBD). We have a translation compiler ("trac") which supports old-style TRS files, but ags4 now writes TRS in PO format. Regardless of whether we keep original "trac" in ags4 branch or not, I suggest to give new tool a new name (e.g. "trac2" or "tracpo") so to distinguish them inside a collection of all tools.

Above lists may be amended if more tasks are required.

**Comment by ericoporto** [comment]
In the auxiliary tasks, the bitmap data task was done in https://github.com/adventuregamestudio/ags/pull/2435

**Comment by ericoporto** [comment]
Did a small update of the message on top

**Comment by ericoporto** [comment]
I guess command line tools could be a goal of 3.6.4 meanwhile AGS 4 keeps going following its own reference ticket, and working with these two things in parallel and avoiding adding anything else other than bug fixes in 3 branch?

**Comment by ivan-mogilko** [comment]
Yes this seems right, ags3 tools can be completed in 3.x branch, and we may tag a 3.6.4 release if only just for releasing complete set of build tools. 3.x branch remains for bug fixes, code cleanup, and backwards compatibility updates (running old games).

**Comment by ericoporto** [comment]
Just asking here because I don’t want to create a ticket for asking this. But in ags3, do we need to introduce some flag for ASCII vs utf-8 in any of the tools - I guess not in the ones that have direct access to Game.agf, but perhaps this is a thing necessary for the compiler?

I was pretty much ignoring this before but just wanted to raise this in the roadmap because I kinda don’t know how/where this needs to be taking into account. I guess ags4 is utf-8 only, so there things are nicer.

**Comment by ivan-mogilko** [comment]
> Just asking here because I don’t want to create a ticket for asking this. But in ags3, do we need to introduce some flag for ASCII vs utf-8 in any of the tools - I guess not in the ones that have direct access to Game.agf, but perhaps this is a thing necessary for the compiler?

agf2dta definitely needs to differentiate between ASCII and Unicode mode, because game texts must be converted and saved in either ANSI or UTF-8. DataFileWriter in the Editor has the TextEncoding setting that is used when converting these texts from the memory to a proper serialization format:
https://github.com/adventuregamestudio/ags/blob/a56d5df3fe6dcbaa2cacae0458d5798b9819d322/Editor/AGS.Editor/DataFileWriter.cs#L27-L30

Script compiler need to know this mode, for reading script files either as ANSI or UTF-8 text. I dont remember by heart if there's any difference in parsing syntax there, maybe something related to how character literals are handled. This has to be checked in the compiler code.

**Comment by ericoporto** [comment]
Ah, for the releases of AGS, I have been thinking of making a package that is essentially the tools+compiler+api_header in a zip (or tar.gz for Linux release and perhaps a dmg for macOS), so the tools could be retrieved separately. I guess they will eventually be part of the editor if we want to switch or have both builds along at some point (with tools build marked as experimental at first?).

The api header I have been thinking that for the releases to append some info as comments on the top (like version) and also to make its extension to be .ash, this is easy to do as an mv command. I think the sh extension is too conflicting with shell scripting. Note that if we want, we could package the api header inside agsbuild tool too, if don’t want to have it as a text file for whatever reason.

**Comment by ivan-mogilko** [comment]
Embedding the api header into the tool exe would be a very bad idea, because that makes the tool work with strictly 1 version of the header.

**Comment by ericoporto** [comment]
Ah right and the other alternative too I forgot to list is to use the .scm file format.

I am mostly thinking about if someone edits the api header directly for whatever reason and reports a bug, how it would be possible to figure out that the bug was there, so I was trying to think about ways to discourage such edits.

> the tool work with strictly 1 version of the header

Ah the tools will probably only release all along with a single version. I have been thinking to switch their versioning for something simpler that would append the AGS version to then, because maybe you didn’t intend to change the tool but a change of the Common may have affected and we would need some way to distinguish these different builds.
